File: Patrones de Comportamiento/composite.py
from abc import ABC, abstractmethod

class ComponenteArchivo(ABC):
    @abstractmethod
    def mostrar(self,nivel=0):
        pass

class Archivo(ComponenteArchivo):
    def __init__(self,nombre):
        self.nombre = nombre

    def mostrar(self,nivel=0):
        print(" " * nivel + f"- {self.nombre}")

class Carpeta(ComponenteArchivo):
    def __init__(self,nombre):
        self.nombre = nombre
        self.componentes = []

    def agregar(self,componente):
        self.componentes.append(componente)

    def mostrar(self, nivel = 0):
        print(" " * nivel + f"+ [{self.nombre}]")
        for componente in self.componentes:
            componente.mostrar(nivel + 1)

archivo1 = Archivo("docuemnto.txt")
archivo2 = Archivo("imagen.tjpg")
archivo3 = Archivo("video.mp4")

carpeta1 = Carpeta("Mis Docuemntos")
carpeta1.agregar(archivo1)
carpeta1.agregar(archivo2)

carpeta2 = Carpeta("Videos")
carpeta2.agregar(archivo3)

carpeta_principal = Carpeta("/")
carpeta_principal.agregar(carpeta1)
carpeta_principal.agregar(carpeta2)

carpeta_principal.mostrar()

File: Patrones de Comportamiento/composite.java
import java.util.ArrayList;
import java.util.List;

abstract class ComponenteArchivo {
    public abstract void mostrar(int nivel);
}

class Archivo extends ComponenteArchivo {
    private String nombre;

    public Archivo(String nombre) {
        this.nombre = nombre;
    }

    @Override
    public void mostrar(int nivel) {
        System.out.println(" ".repeat(nivel) + "- " + nombre);
    }
}

class Carpeta extends ComponenteArchivo {
    private String nombre;
    private List<ComponenteArchivo> componentes;

    public Carpeta(String nombre) {
        this.nombre = nombre;
        this.componentes = new ArrayList<>();
    }

    public void agregar(ComponenteArchivo componente) {
        componentes.add(componente);
    }

    @Override
    public void mostrar(int nivel) {
        System.out.println(" ".repeat(nivel) + "+ [" + nombre + "]");
        for (ComponenteArchivo componente : componentes) {
            componente.mostrar(nivel + 1);
        }
    }
}

public class composite {
    public static void main(String[] args) {
        Archivo archivo1 = new Archivo("proyecto_final.java");
        Archivo archivo2 = new Archivo("presentacion.pptx");
        Archivo archivo3 = new Archivo("reporte.pdf");

        Carpeta carpeta1 = new Carpeta("Trabajo");
        carpeta1.agregar(archivo1);
        carpeta1.agregar(archivo2);

        Carpeta carpeta2 = new Carpeta("Documentos");
        carpeta2.agregar(archivo3);

        Carpeta carpetaPrincipal = new Carpeta("/");
        carpetaPrincipal.agregar(carpeta1);
        carpetaPrincipal.agregar(carpeta2);

        carpetaPrincipal.mostrar(0);
    }
}
